tempesta_fw/http.c

```diff
@@ -383,18 +369,636 @@ tfw_http_send_504(TfwHttpReq *req, const char *reason)
+ */
+static void
+tfw_http_req_fwd(TfwSrvConn *srv_conn, TfwHttpReq *req)
+{
+	LIST_HEAD(equeue);
+
+	TFW_DBG2("%s: srv_conn=[%p], req=[%p]\n", __func__, srv_conn, req);
+	BUG_ON(!(TFW_CONN_TYPE(srv_conn) & Conn_Srv));
+
+	spin_lock(&srv_conn->fwd_qlock);
+	list_add_tail(&req->fwd_list, &srv_conn->fwd_queue);
+	srv_conn->qsize++;
+	if (tfw_http_req_is_nip(req))
+		__tfw_http_req_nip_enlist(srv_conn, req);
+	if (tfw_http_conn_on_hold(srv_conn)) {
+		spin_unlock(&srv_conn->fwd_qlock);
+		return;
+	}
+	__tfw_http_conn_fwd_unsent(srv_conn, &equeue);
+	spin_unlock(&srv_conn->fwd_qlock);
+
+	if (!list_empty(&equeue))
+		tfw_http_req_zap_error(&equeue);
+}
+
+/*
+ * Treat a possible non-idempotent request in case of a connection
+ * repair (re-send or re-schedule).
+ *
+ * A non-idempotent request that was forwarded but not responded to
+ * is not re-sent or re-scheduled by default. Configuration option
+ * can be used to have that request re-sent or re-scheduled.
+ *
+ * As forwarding is paused after a non-idempotent request is sent,
+ * there can be only one such request among forwarded requests, and
+ * that's @srv_conn->msg_sent.
+ *
+ * Note: @srv_conn->msg_sent may change in result.
+ */
+static inline void
+tfw_http_conn_treatnip(TfwSrvConn *srv_conn, struct list_head *equeue)
+{
+	TfwServer *srv = (TfwServer *)srv_conn->peer;
+	TfwHttpReq *req_sent = (TfwHttpReq *)srv_conn->msg_sent;
+
+	if (req_sent && tfw_http_req_is_nip(req_sent)
+	    && likely(!(srv->sg->flags & TFW_SRV_RETRY_NIP)))
+	{
+		BUG_ON(list_empty(&req_sent->nip_list));
+		srv_conn->msg_sent =
+			/* list_is_first(&req_sent->fwd_list, fwd_queue); */
+			(srv_conn->fwd_queue.next == &req_sent->fwd_list) ?
+			NULL : (TfwMsg *)list_prev_entry(req_sent, fwd_list);
+		tfw_http_req_move2equeue(srv_conn, req_sent, equeue, 504);
+	}
+}
+
+/*
+ * Re-forward requests in a server connection. Requests that exceed
+ * the set limits are evicted.
+ */
+static TfwHttpReq *
+tfw_http_conn_resend(TfwSrvConn *srv_conn, bool first, struct list_head *equeue)
+{
+	TfwHttpReq *req, *tmp, *req_resent = NULL;
+	TfwServer *srv = (TfwServer *)srv_conn->peer;
+	struct list_head *end, *fwd_queue = &srv_conn->fwd_queue;
+
+	TFW_DBG2("%s: conn=[%p] one_msg=[%s]\n",
+		 __func__, srv_conn, first ? "true" : "false");
+	BUG_ON(!srv_conn->msg_sent);
+	BUG_ON(list_empty(&((TfwHttpReq *)srv_conn->msg_sent)->fwd_list));
+
+	req = list_first_entry(fwd_queue, TfwHttpReq, fwd_list);
+	end = ((TfwHttpReq *)srv_conn->msg_sent)->fwd_list.next;
+
+	/* Similar to list_for_each_entry_safe_from() */
+	for (tmp = list_next_entry(req, fwd_list);
+	     &req->fwd_list != end;
+	     req = tmp, tmp = list_next_entry(tmp, fwd_list))
+	{
+		if (tfw_http_req_evict_timeout(srv_conn, srv, req, equeue))
+			continue;
+		if (tfw_http_req_evict_retries(srv_conn, srv, req, equeue))
+			continue;
+		if (!tfw_http_req_fwd_send(srv_conn, srv, req, equeue))
+			continue;
+		req_resent = req;
+		if (unlikely(first))
```

Comment:
How do you mean - move `first` to `for()` loop's condition? I am not sure I understand.